### PR TITLE
[#IR-328] As a dev, I have added a data cleaning step in the system

### PR DESCRIPTION
**Pull request by Ravi2714**
In user model firstname,lastname,email,timeproviderconfig                
in sprint model title and sprintid               
in retrospective model   title, project name
task keys from time sheet

### CODE REVIEW

File: apps/retrospective/models/sprint.go [hunk 1]
```diff
@@ -2,13 +2,15 @@ package models
...
 	"errors"
-	"github.com/iReflect/reflect-app/config"
-	customErrors "github.com/iReflect/reflect-app/libs"
-	"github.com/iReflect/reflect-app/libs/utils"
 	"log"
 	"strconv"
+	"strings"
 	"time"
 
+	"github.com/iReflect/reflect-app/config"
```
Comment (chahat-arora) [inline apps/retrospective/models/sprint.go]: sort imports properly.

File: apps/retrospective/services/sprint_sync.go [hunk 2]
```diff
@@ -63,9 +65,29 @@ func (service SprintService) SyncSprintData(sprintID string) (err error) {
...
+
+	json.Unmarshal(taskProviderConfig, &taskProviderConfigList)
+	tpConfig := taskProviderConfigList[0]
+	tp := tpConfig.(map[string]interface{})
+
 	for _, sprintMember := range sprint.SprintMembers {
 		var memberTaskKeys []string
 		memberTaskKeys, timeLogs, err = service.GetSprintMemberTimeTrackerData(sprintMember, sprint)
+
+		if tp["type"].(string) == constants.PivotalTracker {
```
Comment (chahat-arora) [inline apps/retrospective/services/sprint_sync.go]: Please, move this code in the `GetSprintMemberTimeTrackerData(...)` method.

File: apps/user/models/user.go [hunk 3]
```diff
@@ -26,6 +39,20 @@ type User struct {
...
 
+//BeforeSave ...
+func (user *User) BeforeSave() {
+	var timeconfig []TimeProviderConfig
+
+	user.FirstName = strings.TrimSpace(user.FirstName)
+	user.LastName = strings.TrimSpace(user.LastName)
+	user.Email = strings.TrimSpace(user.Email)
+
+	json.Unmarshal([]byte(user.TimeProviderConfig), &timeconfig)
```
Comment (chahat-arora) [inline apps/user/models/user.go]: Please, handle the error case also.

File: apps/user/models/user.go [hunk 4]
```diff
@@ -26,6 +39,20 @@ type User struct {
...
+	timeconfig[0].Data.Email = strings.TrimSpace(timeconfig[0].Data.Email)
+	timeconfig[0].Type = strings.TrimSpace(timeconfig[0].Type)
+	user.TimeProviderConfig, _ = json.Marshal(timeconfig)
```
Comment (chahat-arora) [inline apps/user/models/user.go]: Please, handle the error case also.

File: constants/constants.go [hunk 5]
```diff
@@ -8,6 +8,11 @@ const CustomDateFormat = "2006-01-02"
 // ActionItemType is datatype for ActionItems's
 type ActionItemType string
 
+// constats to be used for task providers
+const (
+	PivotalTracker = "pivotal"
```
Comment (chahat-arora) [inline constants/constants.go]: Please, replace `"pivotal"` string from the whole code by this constant.

File: apps/user/models/user.go [hunk 6]
```diff
@@ -26,6 +39,20 @@ type User struct {
...
+//BeforeSave ...
+func (user *User) BeforeSave() {
+	var timeconfig []TimeProviderConfig
+
+	user.FirstName = strings.TrimSpace(user.FirstName)
+	user.LastName = strings.TrimSpace(user.LastName)
+	user.Email = strings.TrimSpace(user.Email)
+
+	json.Unmarshal([]byte(user.TimeProviderConfig), &timeconfig)
+	timeconfig[0].Data.Email = strings.TrimSpace(timeconfig[0].Data.Email)
```
Comment (chahat-arora) [inline apps/user/models/user.go]: do this for all time providers instead of first.

File: apps/user/models/user.go [hunk 7]
```diff
@@ -26,6 +39,20 @@ type User struct {
 	Profiles           []UserProfile
 }
 
+//BeforeSave ...
+func (user *User) BeforeSave() {
+	var timeconfig []TimeProviderConfig
+
+	user.FirstName = strings.TrimSpace(user.FirstName)
```
Comment (chahat-arora) [inline apps/user/models/user.go:47]: Please, make a model method for this and Also, look for other places.

File: apps/retrospective/services/sprint_sync.go [hunk 8]
```diff
@@ -3,19 +3,21 @@ package services
...
-	"github.com/deckarep/golang-set"
+	mapset "github.com/deckarep/golang-set"
 
 	"github.com/gocraft/work"
 	retroModels "github.com/iReflect/reflect-app/apps/retrospective/models"
 	"github.com/iReflect/reflect-app/apps/tasktracker"
 	taskTrackerSerializers "github.com/iReflect/reflect-app/apps/tasktracker/serializers"
 	"github.com/iReflect/reflect-app/apps/timetracker"
 	timeTrackerSerializers "github.com/iReflect/reflect-app/apps/timetracker/serializers"
+	"github.com/iReflect/reflect-app/constants"
```
Comment (chahat-arora) [inline apps/retrospective/services/sprint_sync.go]: Please, sort imports properly.

**Review by chahat-arora — CHANGES_REQUESTED** [review]
1. Please mention the list of fields for cleaning( both in Github and JIRA).
2. Remove dump.rdb file.

File: apps/retrospective/models/sprint.go [hunk 9]
```diff
@@ -146,6 +148,7 @@ func (sprint *Sprint) validateDateContinuity(baseQuery *gorm.DB, statuses []Spri
 
 // BeforeSave ...
 func (sprint *Sprint) BeforeSave(db *gorm.DB) (err error) {
+	sprint.SprintID = strings.TrimSpace(sprint.SprintID)
```
Comment (himanshu-garg-165) [inline apps/retrospective/models/sprint.go:150]: Shouldn't we do the same for sprint title?

File: apps/retrospective/services/sprint_sync.go (hunk 2 above)
Comment (himanshu-garg-165) [inline apps/retrospective/services/sprint_sync.go]: Move this logic to Pivotal tracker providers code.

File: apps/retrospective/services/sprint_sync.go [hunk 10]
```diff
@@ -218,7 +238,25 @@ func (service SprintService) SyncSprintMemberData(sprintMemberID string) (err er
 		service.SetSyncFailed(sprint.ID)
 		return err
 	}
+	var taskProviderConfigList []interface{}
 
+	json.Unmarshal(taskProviderConfig, &taskProviderConfigList)
+	tpConfig := taskProviderConfigList[0]
+	tp := tpConfig.(map[string]interface{})
+
+	if tp["type"].(string) == constants.PivotalTracker {
```
Comment (himanshu-garg-165) [inline apps/retrospective/services/sprint_sync.go]: Same here

File: apps/retrospective/services/sprint_sync.go [hunk 11]
```diff
@@ -218,7 +238,25 @@ func (service SprintService) SyncSprintMemberData(sprintMemberID string) (err er
 		service.SetSyncFailed(sprint.ID)
 		return err
 	}
+	var taskProviderConfigList []interface{}
 
+	json.Unmarshal(taskProviderConfig, &taskProviderConfigList)
```
Comment (himanshu-garg-165) [inline apps/retrospective/services/sprint_sync.go]: Why are we not handling errors here?

File: apps/user/models/user.go [hunk 12]
```diff
@@ -12,6 +14,17 @@ import (
...
 )
 
+//Email ...
+type Email struct {
+	Email string `json:"email"`
+}
+
+//TimeProviderConfig ...
+type TimeProviderConfig struct {
+	Data Email  `json:"data"`
```
Comment (himanshu-garg-165) [inline apps/user/models/user.go]: This should be an interface and we should handle the type of data dynamically based on the type field.
```suggestion
	Data interface{}  `json:"data"`
```

File: apps/user/models/user.go (hunk 6 above)
Comment (himanshu-garg-165) [inline apps/user/models/user.go]: Update this accordingly.

File: constants/constants.go [hunk 13]
```diff
@@ -8,6 +8,11 @@ const CustomDateFormat = "2006-01-02"
 // ActionItemType is datatype for ActionItems's
 type ActionItemType string
 
+// constats to be used for task providers
```
Comment (himanshu-garg-165) [inline constants/constants.go]: Fix the typo:
```suggestion
// constants to be used for task providers
```

File: constants/constants.go (hunk 5 above)
Comment (himanshu-garg-165) [inline constants/constants.go]: We have this constant in the Pivotal tracker task provider.

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/retrospective/models/sprint.go (hunk 1 above)
Comment (Ravi2714) [reply apps/retrospective/models/sprint.go]: done

File: apps/retrospective/services/sprint_sync.go (hunk 2 above)
Comment (Ravi2714) [reply apps/retrospective/services/sprint_sync.go]: changed

File: apps/user/models/user.go (hunk 3 above)
Comment (Ravi2714) [reply apps/user/models/user.go]: done

File: apps/user/models/user.go (hunk 4 above)
Comment (Ravi2714) [reply apps/user/models/user.go]: done

File: constants/constants.go (hunk 5 above)
Comment (Ravi2714) [reply constants/constants.go]: done

File: apps/user/models/user.go (hunk 6 above)
Comment (Ravi2714) [reply apps/user/models/user.go]: done

File: apps/user/models/user.go (hunk 7 above)
Comment (Ravi2714) [reply apps/user/models/user.go:47]: done

File: apps/retrospective/services/sprint_sync.go (hunk 8 above)
Comment (Ravi2714) [reply apps/retrospective/services/sprint_sync.go]: sorted

File: apps/retrospective/models/sprint.go (hunk 9 above)
Comment (Ravi2714) [reply apps/retrospective/models/sprint.go:150]: done

File: apps/retrospective/services/sprint_sync.go (hunk 2 above)
Comment (Ravi2714) [reply apps/retrospective/services/sprint_sync.go]: done

File: apps/retrospective/services/sprint_sync.go (hunk 10 above)
Comment (Ravi2714) [reply apps/retrospective/services/sprint_sync.go]: - changed

File: apps/retrospective/services/sprint_sync.go (hunk 11 above)
Comment (Ravi2714) [reply apps/retrospective/services/sprint_sync.go]: removed

File: apps/user/models/user.go (hunk 12 above)
Comment (Ravi2714) [reply apps/user/models/user.go]: changed

File: apps/user/models/user.go (hunk 6 above)
Comment (Ravi2714) [reply apps/user/models/user.go]: done

File: constants/constants.go (hunk 13 above)
Comment (Ravi2714) [reply constants/constants.go]: removed

File: constants/constants.go (hunk 5 above)
Comment (Ravi2714) [reply constants/constants.go]: removed

File: apps/timetracker/providers.go [hunk 14]
```diff
@@ -1,9 +1,11 @@
 package timetracker
 
 import (
+	"strings"
 	"time"
```
Comment (chahat-arora) [inline apps/timetracker/providers.go]: Please, sort imports properly.

File: apps/timetracker/providers/gsheet.go [hunk 15]
```diff
@@ -1,12 +1,14 @@
 package providers
 
 import (
-	"github.com/iReflect/reflect-app/libs/utils"
 	"log"
 	"time"
 
+	"github.com/iReflect/reflect-app/libs/utils"
```
Comment (chahat-arora) [inline apps/timetracker/providers/gsheet.go]: same here.

**Review by chahat-arora — CHANGES_REQUESTED** [review]

File: apps/timetracker/providers.go (hunk 14 above)
Comment (Ravi2714) [reply apps/timetracker/providers.go]: done

File: apps/timetracker/providers/gsheet.go (hunk 15 above)
Comment (Ravi2714) [reply apps/timetracker/providers/gsheet.go]: done

File: apps/tasktracker/providers/pivotal.go [hunk 16]
```diff
@@ -153,6 +154,7 @@ func (c *PivotalConnection) serializeTickets(tickets []*pivotal.Story, userIDNam
 // serializeTicket ...
 func (c *PivotalConnection) serializeTicket(ticket *pivotal.Story, userIDNameMap map[int]string) *serializers.Task {
 	ticketID := strconv.Itoa(ticket.Id)
+	ticketID = strings.TrimPrefix(ticketID, "#")
```
Comment (chahat-arora) [inline apps/tasktracker/providers/pivotal.go]: does this really need here?

File: apps/timetracker/providers.go (hunk 14 above)
Comment (chahat-arora) [reply apps/timetracker/providers.go]: check once more.

File: apps/timetracker/providers/gsheet.go [hunk 17]
```diff
@@ -1,17 +1,18 @@
 package providers
 
 import (
-	"github.com/iReflect/reflect-app/libs/utils"
 	"log"
 	"time"
```
Comment (chahat-arora) [inline apps/timetracker/providers/gsheet.go]: sort imports properly.

File: apps/user/models/user.go (hunk 7 above)
Comment (chahat-arora) [reply apps/user/models/user.go:47]: That's not a model method please check once more.

**Review by chahat-arora — CHANGES_REQUESTED** [review]
Please, back merge develop.

File: apps/user/models/user.go (hunk 7 above)
Comment (Ravi2714) [reply apps/user/models/user.go:47]: done

File: apps/timetracker/providers/gsheet.go (hunk 17 above)
Comment (Ravi2714) [reply apps/timetracker/providers/gsheet.go]: done

File: apps/timetracker/providers.go [hunk 18]
```diff
@@ -65,6 +68,12 @@ func GetConnections(config []byte) (connections []Connection, err error) {
 			connections = append(connections, connection)
 		}
 	}
-
 	return connections, nil
 }
+
+// CleanTickets ...
```
Comment (himanshu-garg-165) [inline apps/timetracker/providers.go]: This function would be called for all the trackers, which is not needed, please check if we can have separate tickets sanitization methods.

File: apps/timetracker/providers.go [hunk 19]
```diff
@@ -65,6 +68,12 @@ func GetConnections(config []byte) (connections []Connection, err error) {
...
+func CleanTickets(timeLogs []serializers.TimeLog) {
+	for index, value := range timeLogs {
+		timeLogs[index].TaskKey = strings.TrimPrefix(value.TaskKey, "#")
```
Comment (himanshu-garg-165) [inline apps/timetracker/providers.go]: This is only Pivotal Tracker specific.

File: apps/timetracker/providers/gsheet.go [hunk 20]
```diff
@@ -1,17 +1,18 @@
...
```
Comment (himanshu-garg-165) [inline apps/timetracker/providers/gsheet.go]: Imports are not ordered properly.

File: apps/user/models/user.go [hunk 21]
```diff
@@ -26,6 +34,31 @@ type User struct {
...
+
+	user.FirstName = strings.TrimSpace(user.FirstName)
+	user.LastName = strings.TrimSpace(user.LastName)
+	user.Email = strings.TrimSpace(user.Email)
+
+	err := json.Unmarshal([]byte(user.TimeProviderConfig), &timeProviderConfigurations)
+	if err != nil {
+		return err
+	}
+	for index, timeconfig := range timeProviderConfigurations {
```
Comment (himanshu-garg-165) [inline apps/user/models/user.go]: What if different time provider config has a different format?

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/tasktracker/providers/pivotal.go (hunk 16 above)
Comment (Ravi2714) [reply apps/tasktracker/providers/pivotal.go]: removed

File: apps/timetracker/providers/gsheet.go (hunk 17 above)
Comment (Ravi2714) [reply apps/timetracker/providers/gsheet.go]: done

File: apps/timetracker/providers.go (hunk 18 above)
Comment (Ravi2714) [reply apps/timetracker/providers.go]: done

File: apps/timetracker/providers.go (hunk 19 above)
Comment (Ravi2714) [reply apps/timetracker/providers.go]: done

File: apps/timetracker/providers/gsheet.go (hunk 20 above)
Comment (Ravi2714) [reply apps/timetracker/providers/gsheet.go]: done

File: apps/user/models/user.go (hunk 21 above)
Comment (Ravi2714) [reply apps/user/models/user.go]: done

File: apps/tasktracker/providers/pivotal.go [hunk 22]
```diff
@@ -183,12 +184,20 @@ func (c *PivotalConnection) GetTaskUrl(ticketKey string) string {
...
+	}
+}
+
 // GetTaskList ...
 func (c *PivotalConnection) GetTaskList(ticketKeys []string) []serializers.Task {
 	if len(ticketKeys) == 0 {
 		return nil
 	}
-
+	// To remove # from starting of all ticketkeys if present
```
Comment (himanshu-garg-165) [inline apps/tasktracker/providers/pivotal.go]: Please place this comment in the `CleanTickets` method itself.

File: apps/tasktracker/providers/pivotal.go [hunk 23]
```diff
@@ -183,12 +184,20 @@ func (c *PivotalConnection) GetTaskUrl(ticketKey string) string {
 	return fmt.Sprintf("https://www.pivotaltracker.com/story/show/%v", ticketKey)
 }
 
+// CleanTickets ...
+func (c *PivotalConnection) CleanTickets(ticketKeys []string) {
```
Comment (himanshu-garg-165) [inline apps/tasktracker/providers/pivotal.go]: If it is an internal method then we should make the first letter small, i.e., `cleanTickets`.

File: apps/user/models/user.go [hunk 24]
```diff
@@ -26,6 +34,33 @@ type User struct {
...
+	user.LastName = strings.TrimSpace(user.LastName)
+	user.Email = strings.TrimSpace(user.Email)
+
+	err := json.Unmarshal([]byte(user.TimeProviderConfig), &timeProviderConfigurations)
+	if err != nil {
+		return err
+	}
+	for _, timeConfig := range timeProviderConfigurations {
+		timeConfig.Type = strings.TrimSpace(timeConfig.Type)
+		if timeConfig.Type == "gsheet" {
```
Comment (himanshu-garg-165) [inline apps/user/models/user.go]: Please update it as discussed.

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/tasktracker/providers/pivotal.go (hunk 22 above)
Comment (Ravi2714) [reply apps/tasktracker/providers/pivotal.go]: done

File: apps/tasktracker/providers/pivotal.go (hunk 23 above)
Comment (Ravi2714) [reply apps/tasktracker/providers/pivotal.go]: done

File: apps/user/models/user.go (hunk 24 above)
Comment (Ravi2714) [reply apps/user/models/user.go]: done

File: apps/tasktracker/providers/pivotal.go [hunk 25]
```diff
@@ -183,12 +184,20 @@ func (c *PivotalConnection) GetTaskUrl(ticketKey string) string {
 	return fmt.Sprintf("https://www.pivotaltracker.com/story/show/%v", ticketKey)
 }
 
+// cleanTickets ...
+func (c *PivotalConnection) cleanTickets(ticketKey string) string {
```
Comment (himanshu-garg-165) [inline apps/tasktracker/providers/pivotal.go]:
```suggestion
func (c *PivotalConnection) sanitizeTicket(ticketKey string) string {
```

File: apps/timetracker/providers/gsheet.go [hunk 26]
```diff
@@ -88,6 +89,12 @@ func getConfigObject(config interface{}) (GsheetConfig, error) {
 	return c, nil
 }
 
+// CleanTimeProviderConfig ...
+func (m *GsheetConnection) CleanTimeProviderConfig(data interface{}) interface{} {
+	data.(map[string]interface{})["email"] = strings.TrimSpace(data.(map[string]interface{})["email"].(string))
```
Comment (himanshu-garg-165) [inline apps/timetracker/providers/gsheet.go]: We should use the connection's config object since it is being parsed when the connection object is created.

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/timetracker/providers/gsheet.go (hunk 26 above)
Comment (Ravi2714) [reply apps/timetracker/providers/gsheet.go]: done

File: apps/tasktracker/providers/pivotal.go (hunk 25 above)
Comment (Ravi2714) [reply apps/tasktracker/providers/pivotal.go]: done

**Review by himanshu-garg-165 — APPROVED** [review]

**Review by chahat-arora — APPROVED** [review]

**Review by chahat-arora — APPROVED** [review]

**Pull request merged**